### PR TITLE
[FEATURE: UIPT-26] - Product order: Product details

### PR DESCRIPTION
**Pull request by castan1c**

### CODE REVIEW

File: src/app/components/product-order/product-order.component.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
+import { Component, Input, OnInit } from '@angular/core';
+
+import { IProduct } from 'src/app/interfaces/product.interface';
+import { IProductDetails } from 'src/app/interfaces/product-details.interface';
+import { faHeart, IconDefinition } from '@fortawesome/free-solid-svg-icons';
```
Comment (AlexLarusso) [inline src/app/components/product-order/product-order.component.ts]: Move after all imported interfaces. 
Try to sort imports.

File: src/app/components/product-order/product-order.component.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  public handleColorSelect(event: number): void {
+    this.selectedColor = event;
+  }
+
+  public handleQuantitySelect(event: number): void {
+    this.selectedQty = event;
+  }
+
+  public onBuyClick(): void {
+    // TODO: imlement after CartComponent is implemented
```
Comment (AlexLarusso) [inline src/app/components/product-order/product-order.component.ts]: Typos -> `imPlement`

File: src/app/components/product-order/product-order.component.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    const message = `Added ${this.productDetails.title} to your cart.
+      Quantity: ${this.selectedQty}.
+      Size: ${this.selectedSize}.
+      Color: ${this.selectedColor}.
+      Full price: ${this.selectedQty * this.productDetails.price} uah.`
+
+    alert(message);
```
Comment (AlexLarusso) [inline src/app/components/product-order/product-order.component.ts]: TODO: Implement modal service

File: src/app/components/product-order/product-order.component.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  public addToWhishlistClick(): void {
+    // TODO: imlement after WhishlistComponent is implemented
+    alert(`${this.productDetails.title} added  to whishlist`);
```
Comment (AlexLarusso) [inline src/app/components/product-order/product-order.component.ts]: Same

File: src/app/components/product-preview/product-preview.component.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import { IProductImage } from 'src/app/interfaces/product-image.interface';
+
+@Component({
+  selector: 'app-product-preview',
+  templateUrl: './product-preview.html',
+  styleUrls: ['./product-preview.scss']
+})
+export class ProductPreviewComponent implements OnChanges {
+  @Input() public productImages: Array<IProductImage>;
+  @Input() public productColorSelected: number;
```
Comment (AlexLarusso) [inline src/app/components/product-preview/product-preview.component.ts]: `productColorSelected` -> please name it more clearly. 
Maybe `productColorIdSelected` or smth like this.

File: src/app/components/product-preview/product-preview.component.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+  public imagesSource: Array<string>;
+  public selectedImageID: number;
+
+  public ngOnChanges(): void {
+    this.selectedImageID = 0;
+    if(this.productColorSelected) {
```
Comment (AlexLarusso) [inline src/app/components/product-preview/product-preview.component.ts]: space after `if`

File: src/app/components/product-preview/product-preview.component.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+      this.imagesSource = this.productImages.find(images =>
```
Comment (AlexLarusso) [inline src/app/components/product-preview/product-preview.component.ts]: I am not sure that `find` is the correct function in this case. 
Please explain or use another one.

File: src/app/components/product-preview/product-preview.component.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        images.value == this.productColorSelected).url;
+    } else {
+      this.imagesSource = this.productImages[0].url;
+    }
+  }
+
+  public onImageSelect(i: number): void {
```
Comment (AlexLarusso) [inline src/app/components/product-preview/product-preview.component.ts]: `i` more often means index. Is this right in your case?

File: src/app/components/product-preview/product-preview.html [hunk 9]
```diff
@@ -0,0 +1,11 @@
+<div *ngIf="imagesSource.length" class="product-preview">
```
Comment (AlexLarusso) [inline src/app/components/product-preview/product-preview.html]: `imagesSource?.length`

File: src/app/pages/product-details-page/product-details-page.component.ts [hunk 10]
```diff
@@ -1,11 +1,45 @@
...
+    private shortInfoService: ProductShortInfoService,
+    private route: ActivatedRoute) { }
+
+  public ngOnInit(): void {
+    this.route.params.subscribe(value => {
+      this.productSource$ = this.productService
+        .getProduct(value.id);
+    });
+
+    this.productService
```
Comment (AlexLarusso) [inline src/app/pages/product-details-page/product-details-page.component.ts]: `this.productSub = `

File: src/app/shared/accordeon/accordeon.component.ts [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+export class AccordeonComponent implements OnInit {
+  @Input() content: { title: string; text: string; }[];
+
+  public selectedTopic: number;
+  public iconExpand: IconDefinition = faPlus;
+  public iconCollapse: IconDefinition = faMinus;
+
+  private collapsedAllIndex: number = -1;
+
+  constructor() { }
```
Comment (AlexLarusso) [inline src/app/shared/accordeon/accordeon.component.ts]: Please remove constructor if you don't use it

File: src/app/shared/accordeon/accordeon.component.ts [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+  public ngOnInit() {
```
Comment (AlexLarusso) [inline src/app/shared/accordeon/accordeon.component.ts]: Don't forget about returned type: void here.

File: src/app/shared/accordeon/accordeon.component.ts [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    this.selectedTopic = this.collapsedAllIndex;
+  }
+
+  public topicClick(i: number) {
```
Comment (AlexLarusso) [inline src/app/shared/accordeon/accordeon.component.ts]: same

File: src/app/shared/dropdown/dropdown-component.ts [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+@Component({
+  selector: 'app-dropdown',
+  templateUrl: './dropdown.html',
+  styleUrls: ['./dropdown.scss']
+})
+export class DropdownComponent implements OnInit {
+  @Input() options: Array<string>;
+  @Input() default: string;
+  @Output() itemSelect: EventEmitter<string> = new EventEmitter<string>();
```
Comment (AlexLarusso) [inline src/app/shared/dropdown/dropdown-component.ts]: empty space before

File: src/app/shared/dropdown/dropdown-component.ts [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+  
+  private iconUp: IconDefinition;
+  private iconDown: IconDefinition;
+  private selected: string;
+  private expanded: boolean;
+
+  constructor() { }
```
Comment (AlexLarusso) [inline src/app/shared/dropdown/dropdown-component.ts]: Remove useless constructor

File: src/app/shared/dropdown/dropdown-component.ts [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+
+  ngOnInit() {
```
Comment (AlexLarusso) [inline src/app/shared/dropdown/dropdown-component.ts]: Typescript! Please check everywhere

File: src/app/shared/dropdown/dropdown.html [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+  (blur)="expanded=false">
+  <span class="dropdown__selected">{{ selected }}</span>
+  <ul class="options">
+    <li *ngFor="let option of options; let i = index"
+      class="options__item"
+      (click)="onItemSelect(i)">
+      <span>{{ option }}</span>
+    </li>
+  </ul>
+  <fa-icon class="icon" [icon]="expanded?iconUp:iconDown"></fa-icon>
```
Comment (AlexLarusso) [inline src/app/shared/dropdown/dropdown.html]: `expanded?iconUp:iconDown` -> spaces

File: src/app/shared/dropdown/dropdown-component.ts [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+
+  onBlur() {
+    this.expanded = false;
+  }
+
+  onItemSelect(i: number) {
+    this.selected = this.options[i];
+    this.itemSelect.emit(this.selected);
+  }
+}
```
Comment (AlexLarusso) [inline src/app/shared/dropdown/dropdown-component.ts]: Is this you code?

File: src/app/shared/select-color/select-color.html [hunk 19]
```diff
@@ -0,0 +1,12 @@
+<div class="select-color__wrapper">
+  <ng-content></ng-content>
+  <div class="select-color">
+    <button *ngFor="let color of colors; let i = index"
+      class="colorBtn"
+      [class.selected]="selectedColorID === i"
+      [ngStyle]="{backgroundColor: color}"
```
Comment (AlexLarusso) [inline src/app/shared/select-color/select-color.html]: `{backgroundColor: color}` -> add spaces

File: src/app/shared/select-number/select-number.component.ts [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+  @Output() numberSelect: EventEmitter<number> = new EventEmitter<number>();
+
+  public minusIcon: IconDefinition;
+  public plusIcon: IconDefinition;
+  public value: number;
+  public valueLimit: boolean;
+  public increment = 1;
+  public decrement = -1;
+
+  constructor() { }
```
Comment (AlexLarusso) [inline src/app/shared/select-number/select-number.component.ts]: remove

File: src/app/shared/select-number/select-number.component.ts [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+
+  ngOnInit() {
```
Comment (AlexLarusso) [inline src/app/shared/select-number/select-number.component.ts]: typescript

File: src/app/shared/select-number/select-number.component.ts [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+    this.minusIcon = faMinusCircle;
+    this.plusIcon = faPlusCircle;
+    this.value = 1;
+    this.valueLimit = false;
+  }
+
+  onClick(i: number) {
+    const nextNum = this.value + i
+    if (nextNum <= this.maxNumber && nextNum > 0) {
```
Comment (AlexLarusso) [inline src/app/shared/select-number/select-number.component.ts]: empty space before

File: src/app/shared/select-number/select-number.component.ts [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+
+  ngOnInit() {
+    this.minusIcon = faMinusCircle;
+    this.plusIcon = faPlusCircle;
+    this.value = 1;
+    this.valueLimit = false;
+  }
+
+  onClick(i: number) {
+    const nextNum = this.value + i
```
Comment (AlexLarusso) [inline src/app/shared/select-number/select-number.component.ts]: semicolon

File: src/app/shared/select-number/select-number.component.ts [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+    if (nextNum <= this.maxNumber && nextNum > 0) {
+        this.value = nextNum;
+        this.numberSelect.emit(nextNum);
+        this.valueLimit = false;
+    } else {
+        this.valueLimit = true;
+        setTimeout(() => {
```
Comment (AlexLarusso) [inline src/app/shared/select-number/select-number.component.ts]: rxjs timeout, or detectChanges

File: src/app/shared/services/product.service.ts [hunk 25]
```diff
@@ -13,6 +13,8 @@ export class ProductService {
 
   public getProduct(id): Observable<IProduct> {
     return this.httpService.getData().pipe(map(products =>
-      products.find(product => product.id === id)));
+      products.find(product => {
+        return product.id == id;
```
Comment (AlexLarusso) [inline src/app/shared/services/product.service.ts]: Use strict comparison. One-lined arrow function

File: src/app/components/product-preview/product-preview.component.ts (hunk 5 above)
Comment (castan1c) [reply src/app/components/product-preview/product-preview.component.ts]: its color value, which is a number in our DB:
```"colors": ["#edfcfb", "#fffff", "#debb47"],
    "images": [
      {
        "value": "#edfcfb",
        "url": ["./assets/server-data/images/blouse_5_1.jpg",
          "./assets/server-data/images/blouse_5_2.jpg",
          "./assets/server-data/images/blouse_5_3.jpg",
          "./assets/server-data/images/blouse_5_4.jpg"]
      },
```
I renamed it to productSelectedColor

File: src/app/shared/dropdown/dropdown-component.ts (hunk 18 above)
Comment (castan1c) [reply src/app/shared/dropdown/dropdown-component.ts]: It is!

File: src/app/components/product-details/product-details.component.ts [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+
+import { IProductDetails } from "src/app/interfaces/product-details.interface";
+
+@Component({
+  selector: 'app-product-details',
+  templateUrl: "./product-details.html",
+  styleUrls: ["./product-details.scss"]
+})
+export class ProductDetailsComponent implements OnInit {
+  @Input() productDetails: IProductDetails;
```
Comment (AlexLarusso) [inline src/app/components/product-details/product-details.component.ts]: `public`

File: src/app/components/product-order/product-order.component.ts [hunk 27]
```diff
@@ -0,0 +1,83 @@
...
+const DELIVERY_MOCK = 'Officia sint Lorem do officia velit voluptate. Dolor commodo pariatur irure do excepteur ullamco commodo pariatur et. Esse velit incididunt qui incididunt consectetur ea sit excepteur ex eu. Nisi esse dolore aute laborum.'
+const STYLE_MOCK = 'Ullamco eu ut consequat eu sit nostrud occaecat ad nulla nisi cupidatat.';
+
+@Component({
+  selector: 'app-product-order',
+  templateUrl: './product-order.html',
+  styleUrls: ['./product-order.scss']
+})
+export class ProductOrderComponent implements OnInit {
+  @Input() product: IProduct;
```
Comment (AlexLarusso) [inline src/app/components/product-order/product-order.component.ts]: `private` or `public` ?

File: src/app/components/product-order/product-order.component.ts [hunk 28]
```diff
@@ -0,0 +1,83 @@
...
+      category: this.product.category,
+      sex: this.product.sex,
+      season: this.product.season,
+      options: productOptions,
+      description: productDescription,
+    }
+    this.productImages = this.product.images;
+  }
+
+  public handleSizeSelect(event: string): void {
```
Comment (AlexLarusso) [inline src/app/components/product-order/product-order.component.ts]: `event` -> Event couldn't be a string. So please rename argument

Check everywhere

File: src/app/components/product-order/product-order.html [hunk 29]
```diff
@@ -0,0 +1,34 @@
+<div *ngIf="productDetails" class="product-order content__container">
+  <div class="product-order__image-preview">
+    <app-product-preview
+      [productImages]="productImages"
+      [productSelectedColor]="selectedColor">
+    </app-product-preview>
+  </div>
+  <app-product-details [productDetails]="productDetails">
+    <div class="product-options">
+      <app-dropdown class="option" [default]="'Size'" [options]="productDetails.options.sizes"
```
Comment (AlexLarusso) [inline src/app/components/product-order/product-order.html]: Don't use binding if you have static value:
`[default]="'Size'"` -> default="Size"

File: src/app/components/product-preview/product-preview.component.ts [hunk 30]
```diff
@@ -0,0 +1,31 @@
...
+  }
+
+  public onImageSelect(index: number): void {
+    this.selectedImageID = index;
+  }
+
+  private findImagesSource(): Array<string> {
+    return this.productSelectedColor
+      ? this.productImages
+          .find(images => images.value === this.productSelectedColor).url
```
Comment (AlexLarusso) [inline src/app/components/product-preview/product-preview.component.ts]: Forget about one thing:

`this.productImages.find(images => images.value === this.productSelectedColor).url`
undefined.url // Uncaught TypeError: Cannot read property 'url' of undefined

File: src/app/components/product-preview/product-preview.html [hunk 31]
```diff
@@ -0,0 +1,11 @@
+<div *ngIf="imagesSource?.length" class="product-preview">
+  <img class="img_preview"
+    [src]="imagesSource[selectedImageID]"
+    alt="{{ title }}"
+    title="{{ title }}">
```
Comment (AlexLarusso) [inline src/app/components/product-preview/product-preview.html]: Please sort:
```
<img 
    class="img_preview"
    alt="{{ title }}"
    title="{{ title }}"
    [src]="imagesSource[selectedImageID]"
>
```

First should go simple attributes, then bindings

File: src/app/pages/product-details-page/product-details-page.component.ts [hunk 32]
```diff
@@ -1,11 +1,46 @@
...
+  public productsimilarOptionsSub: Subscription;
+
+  constructor(private productService: ProductService,
+    private shortInfoService: ProductShortInfoService,
+    private route: ActivatedRoute) { }
+
+  public ngOnInit(): void {
+    this.productSub = this.route.params
+      .subscribe(value => this.productSource$ =
+        this.productService.getProduct(value.id));
```
Comment (AlexLarusso) [inline src/app/pages/product-details-page/product-details-page.component.ts]: Please have a look to `switchMap`

File: src/app/pages/product-details-page/product-details-page.html [hunk 33]
```diff
@@ -1 +1,3 @@
-<app-product-details></app-product-details>
+<app-product-order *ngIf="productSource$ | async as product"
+  [product]="product"></app-product-order>
```
Comment (AlexLarusso) [inline src/app/pages/product-details-page/product-details-page.html]: Please align

File: src/app/shared/services/product.service.ts [hunk 34]
```diff
@@ -13,6 +13,7 @@ export class ProductService {
 
   public getProduct(id): Observable<IProduct> {
     return this.httpService.getData().pipe(map(products =>
```
Comment (AlexLarusso) [inline src/app/shared/services/product.service.ts]: use `rxjs operator find` instead of `map`

File: src/app/shared/services/product.service.ts [hunk 35]
```diff
@@ -13,6 +13,7 @@ export class ProductService {
 
   public getProduct(id): Observable<IProduct> {
```
Comment (AlexLarusso) [inline src/app/shared/services/product.service.ts]: `getProduct(id: number)`

Comment (castan1c) [reply src/app/shared/services/product.service.ts]: left string | number for now, will be fixed in Refactoring branch

File: src/app/shared/services/product.service.ts (hunk 34 above)
Comment (castan1c) [reply src/app/shared/services/product.service.ts]: I've tried this: ` return this.httpService.getData().pipe(
      find(product => product.id === id)
    );` but it doesn't work as expected...

**Pull request merged**